modules/fundamental/src/analysis/service/mod.rs

```diff
@@ -0,0 +1,584 @@
+
+        // load package relationships into graph
+        for row in relationship_results {
+            let sbom_published = row
+                .try_get("", "published")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            let document_id = row
+                .try_get("", "document_id")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            let product_name = row
+                .try_get("", "product_name")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            let product_version = row
+                .try_get("", "product_version")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+
+            let left_purl_string: String = row
+                .try_get("", "left_qualified_purl")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            if left_purl_string == "NOVALUE" {
+                log::warn!("sbom:{} had packages with no pURL.", document_id);
+                continue;
+            }
+
+            let left_purl: Purl = Purl::from_str(left_purl_string.as_str())?;
+            let p1 = add_node(
+                &mut g,
+                &mut nodes,
+                &left_purl_string,
+                &left_purl.name,
+                &sbom_published,
+                &document_id,
+                &product_name,
+                &product_version,
+            );
+
+            let relationship: Relationship = row.try_get("", "relationship")?;
+            let right_purl_string: String = row
+                .try_get("", "right_qualified_purl")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            if right_purl_string == "NOVALUE" {
+                log::warn!("sbom:{} had packages with no pURL.", document_id);
+                continue;
+            }
+            let right_purl: Purl = Purl::from_str(right_purl_string.as_str())?;
+            let p2 = add_node(
+                &mut g,
+                &mut nodes,
+                &right_purl_string,
+                &right_purl.name,
+                &sbom_published,
+                &document_id,
+                &product_name,
+                &product_version,
+            );
+
+            g.add_edge(p1, p2, relationship);
+        }
+
+        let mut components: Vec<AdvisoryGraphSummary> = Vec::new();
+
+        for node_index in regex_purl_find(&nodes, query.q.as_str()) {
+            if let Some(find_match_package_node) = g.node_weight(node_index) {
+                let mut ancestor_nodes: Vec<NodeIndex> = ancestor_nodes(&g, node_index, node_index)
+                    .into_iter()
+                    .collect();
+
+                ancestor_nodes.sort();
+
+                let mut root_components: Vec<PackageNode> = ancestor_nodes
+                    .iter()
+                    .filter_map(|&ancestor_node| g.node_weight(ancestor_node).cloned())
```

Comment:
dunno if worthwhile or useful, but you can elide the `.cloned()` within the closure, and add a `.cloned()` after `.filter_map(...)`